### PR TITLE
docker can use env conf

### PR DESCRIPTION
**Pull request by xxaier**
* docker can use env conf
    * add docker/build.sh for build docker in local
    * add docker-compose.yml

### CODE REVIEW

**Review by PragmaTwice — CHANGES_REQUESTED** [review]
Hi @xxaier, thanks for your contribution.

However, I do not think it is necessary to pass environment variables as config options to kvrocks via some ugly text substitution. If you want to pass some options to kvrocks in docker, you can just try `docker run apache/kvrocks --some-key some-value --other-key other-value ...`.

Besides, we do not prefer to add more shell scripts, which can hardly be general enough.

**Comment by xxaier** [comment]
docker run apache/kvrocks --some-key some-value --other-key other-value ....

will run , but if you write docker-compose.yml and need read env for a .env, how to write ? change cmd ? Such an operation is very unintuitive.

These bash scripts can be run on all platforms with docker and bash installed.

Configuring docker through environment variables I think is a very common practice

for exampe nginx : https://hub.docker.com/_/nginx

image

for example postgres : https://hub.docker.com/_/postgres
image

for example redis : https://hub.docker.com/_/redis
image

**Comment by PragmaTwice** [comment]
I acknowledge your idea to pass env vars as options, but IMHO the shell script composed by `sed` and some other text substitution tools looks terrible to me since it is more like workaround than a real solution.

BTW, shell scripts are hard to maintain and be general-purpose enough. Our development scripts are all written in python, refer to `x.py`.

For [docker/build.sh](https://github.com/apache/incubator-kvrocks/pull/1474/files#diff-efe74abc7213e95246105b4a7b029557ef16ff997e9cc13eadcfffa18e6757c0), I think we do not need the script. You can learn how  our community build and release docker images from https://kvrocks.apache.org/community/create-a-release .

And for [docker/docker-compose.yml](https://github.com/apache/incubator-kvrocks/pull/1474/files#diff-423deb13b7c401b1a7f41ee91c77f722e11d2f317d6a66b546524e8a04cc8b03), I think we do not need an example compose file, and I really do not know why these arguments are chosen and are suitable for various kvrocks users, e.g. `mem_limit: 2G`  `memswap_limit: -1`.

**Comment by xxaier** [comment]
> I acknowledge your idea to pass env vars as options, but IMHO the shell script composed by `sed` and some other text substitution tools looks terrible to me since it is more like workaround than a real solution.
> 
> BTW, shell scripts are hard to maintain and be general-purpose enough. Our development scripts are all written in python, refer to `x.py`.
> 
> For [docker/build.sh](https://github.com/apache/incubator-kvrocks/pull/1474/files#diff-efe74abc7213e95246105b4a7b029557ef16ff997e9cc13eadcfffa18e6757c0), I think we do not need the script. You can learn how our community build and release docker images from https://kvrocks.apache.org/community/create-a-release .
> 
> And for [docker/docker-compose.yml](https://github.com/apache/incubator-kvrocks/pull/1474/files#diff-423deb13b7c401b1a7f41ee91c77f722e11d2f317d6a66b546524e8a04cc8b03), I think we do not need an example compose file, and I really do not know why these arguments are chosen and are suitable for various kvrocks users, e.g. `mem_limit: 2G` `memswap_limit: -1`.

The reason why I don't use python is that in order to keep the size of the final docker container from growing, the final docker image will not contain the python runtime.

The reason I added build.sh is that when developing and debugging docker, there will be a need for local construction, such as building a database based on kvrocks+topling https://github.com/apache/incubator-kvrocks/pull/1206.

I think local development and building docker debugging is also a common scenario, and it is very troublesome to manually enter commands every time.

**Comment by torwig** [comment]
@xxaier Maybe, we can implement a feature to pass options via an env variable similar to `REDIS_ARGS`. However, personally, I don't like the trick with a bash script to 'translate` env variables to Kvrocks options.

**Comment by PragmaTwice** [comment]
> @xxaier Maybe, we can implement a feature to pass options via an env variable similar to `REDIS_ARGS`. However, personally, I don't like the trick with a bash script to 'translate` env variables to Kvrocks options.

+1. `REDIS_ARGS` looks fine to me.

**Comment by xxaier** [comment]
image
I changed code, read command line args from KVROCKS_ARGS

File: Dockerfile [hunk 1]
```diff
@@ -57,10 +57,9 @@ VOLUME /var/lib/kvrocks
...
 
-COPY ./LICENSE ./NOTICE ./DISCLAIMER ./
+COPY ./docker/run.sh ./LICENSE ./NOTICE ./DISCLAIMER ./
 COPY ./licenses ./licenses
 COPY ./kvrocks.conf /var/lib/kvrocks/
-
 EXPOSE 6666:6666
 
-ENTRYPOINT ["./bin/kvrocks", "-c", "/var/lib/kvrocks/kvrocks.conf", "--dir", "/var/lib/kvrocks", "--pidfile", "/var/run/kvrocks/kvrocks.pid"]
+ENTRYPOINT ["./run.sh"]
```
Comment (git-hulk) [inline Dockerfile:65]: Can we only add the KVROCKS_ARGS for the custom configuration and leave the others as they should(remove the run.sh and docker-compose.yml)?

Comment (xxaier) [reply Dockerfile:65]: see https://stackoverflow.com/questions/37904682/how-do-i-use-docker-environment-variable-in-entrypoint-array
hard to do this

Comment (xxaier) [reply Dockerfile:65]: see https://stackoverflow.com/questions/37904682/how-do-i-use-docker-environment-variable-in-entrypoint-array
hard to do this

**Comment by git-hulk** [comment]
Close the PR due to inactive, feel free to reopen if you like.

**Pull request closed without merging**